### ISSUE TITLE
Not able to create Dask client using AzureMLCluster object

### ISSUE THREAD

**Issue by arnabbiswas1** [dask/dask-cloudprovider#145]
I have created a Dask cluster on Azure ML using the following API. 
```
amlcluster = AzureMLCluster(ws,
                            vm_size="STANDARD_D1",
                            datastores=[Datastore.get(ws, "my_datastore")], 
                            environment_definition=ws.environments['AzureML-Dask-CPU'], 
                            initial_node_count=2, 
                            scheduler_idle_timeout=600,
                            vnet='some_vnet',
                            subnet='subnet1',
                            vnet_resource_group='some_rsrc_group',
                            ct_name="my_dask_cluster"
)
```
Once the cluster is created, if I try to print the variable `amlcluster` in Jupyter Lab, it throws the following error. 

> KeyError                                  Traceback (most recent call last)
> /anaconda/envs/azureml_custom_py37/lib/python3.7/site-packages/IPython/core/formatters.py in __call__(self, obj)
>     916             method = get_real_method(obj, self.print_method)
>     917             if method is not None:
> --> 918                 method()
>     919                 return True
>     920 
> 
> /anaconda/envs/azureml_custom_py37/lib/python3.7/site-packages/distributed/deploy/cluster.py in _ipython_display_(self, **kwargs)
>     361             from IPython.display import display
>     362 
> --> 363             data = {"text/plain": repr(self), "text/html": self._repr_html_()}
>     364             display(data, raw=True)
>     365 
> 
> /anaconda/envs/azureml_custom_py37/lib/python3.7/site-packages/distributed/deploy/cluster.py in __repr__(self)
>     389             self._cluster_class_name,
>     390             self.scheduler_address,
> --> 391             len(self.scheduler_info["workers"]),
>     392             sum(w["nthreads"] for w in self.scheduler_info["workers"].values()),
>     393         )
> 
> KeyError: 'workers'

After the error, it provides just a Dashboard Link. Not sure if it is supposed to print anything else.

![image](https://user-images.githubusercontent.com/4428191/96120521-41080180-0f0c-11eb-87f6-989a2555ce50.png)

If I try to create a Dask Client that alos fails:
```
client = Client(amlcluster)
```
This is how the library version looks like for me:

> dask                      2.20.0                     py_0
> dask-cloudprovider        0.4.1                     <pip>
> dask-core                 2.20.0                     py_0
> dask-glm                  0.2.0                      py_1    conda-forge
> dask-ml                   1.6.0                      py_0    conda-forge
> dask-xgboost              0.1.10                    <pip>

**Comment by quasiben** [comment]
@drabastomek would you have time to look at this ?

**Comment by lostmygithubaccount** [comment]
are you seeing this consistently?

**Comment by arnabbiswas1** [comment]
> are you seeing this consistently?

Yes. Consistently.

**Comment by drabastomek** [comment]
This is surprising as the `self.scheduler_info['workers']` is defined and maintained in the super-class and we do not even touch it in our code... [https://github.com/dask/distributed/blob/dcb46d07318282d8a8976aadb0e4003144a42ae7/distributed/deploy/cluster.py#L55](url)

@quasiben I am afraid I won't be able to help much as I no longer have access to test any functionality of AzureML...

**Comment by lostmygithubaccount** [comment]
@arnabbiswas1 are you on the same vnet as the cluster trying to access it? 

@quasiben @drabastomek fyi I setup a test which runs every 2 hours here: https://github.com/Azure/azureml-examples/actions?query=workflow%3Arun-tutorial-ud - it is mostly green, majority of the failures have been my fault

**Comment by arnabbiswas1** [comment]
@lostmygithubaccount Yes, I am starting the cluster from the Jupyter Lab installed on the compute instance (with in AzureML workspace). So, as per my understanding the VNET should be same.

Just to note the Dask cluster gets created, but printing the variable "amlcluster" throws the above error.

**Comment by jacobtomlinson** [comment]
I think `self.scheduler_info` is populated by an RPC. So perhaps there is some connectivity issue with the scheduler. Are you able to connect a client to the cluster?

**Comment by arnabbiswas1** [comment]
@jacobtomlinson you are correct. If I try the following code from the Jupyter Lab, it fails.

`client = Client(amlcluster)`

If I look into the stack trace (pasted at the end of this message), I see the following error message:

`Timed out trying to connect to 'tcp://10.233.0.7:8786' after 10 s`

As you understand, "10.233.0.7" is the private IP for the dask node.

This client code is running on the Jupyter Lab installed on the compute instance (with in AzureML workspace). The compute instance (private IP: 10.233.0.5) and the dask cluster created should be in the same network. Hence, ideally the the code running on the Jupyter Lab should have access to 'tcp://10.233.0.7:8786'. But, I am not sure what is going wrong.

Any suggestion (how to debug etc. will be appreciated). @lostmygithubaccount 


**Detailed stack trace**:

```
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
/anaconda/envs/azureml_custom_py37/lib/python3.7/site-packages/distributed/comm/core.py in connect(addr, timeout, deserialize, **connection_args)
    233             if not comm:
--> 234                 _raise(error)
    235         except FatalCommClosedError:

/anaconda/envs/azureml_custom_py37/lib/python3.7/site-packages/distributed/comm/core.py in _raise(error)
    214         )
--> 215         raise IOError(msg)
    216 

OSError: Timed out trying to connect to 'tcp://10.233.0.7:8786' after 10 s: in <distributed.comm.tcp.TCPConnector object at 0x7faaa01bca50>: ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

OSError                                   Traceback (most recent call last)
<ipython-input-21-d8364f1d6028> in <module>
----> 1 client = Client(amlcluster)

/anaconda/envs/azureml_custom_py37/lib/python3.7/site-packages/distributed/client.py in __init__(self, address, loop, timeout, set_as_default, scheduler_file, security, asynchronous, name, heartbeat_interval, serializers, deserializers, extensions, direct_to_workers, connection_limit, **kwargs)
    742             ext(self)
    743 
--> 744         self.start(timeout=timeout)
    745         Client._instances.add(self)
    746 

/anaconda/envs/azureml_custom_py37/lib/python3.7/site-packages/distributed/client.py in start(self, **kwargs)
    946             self._started = asyncio.ensure_future(self._start(**kwargs))
    947         else:
--> 948             sync(self.loop, self._start, **kwargs)
    949 
    950     def __await__(self):

/anaconda/envs/azureml_custom_py37/lib/python3.7/site-packages/distributed/utils.py in sync(loop, func, callback_timeout, *args, **kwargs)
    337     if error[0]:
    338         typ, exc, tb = error[0]
--> 339         raise exc.with_traceback(tb)
    340     else:
    341         return result[0]

/anaconda/envs/azureml_custom_py37/lib/python3.7/site-packages/distributed/utils.py in f()
    321             if callback_timeout is not None:
    322                 future = asyncio.wait_for(future, callback_timeout)
--> 323             result[0] = yield future
    324         except Exception as exc:
    325             error[0] = sys.exc_info()

/anaconda/envs/azureml_custom_py37/lib/python3.7/site-packages/tornado/gen.py in run(self)
    733 
    734                     try:
--> 735                         value = future.result()
    736                     except Exception:
    737                         exc_info = sys.exc_info()

/anaconda/envs/azureml_custom_py37/lib/python3.7/site-packages/distributed/client.py in _start(self, timeout, **kwargs)
   1043 
   1044         try:
-> 1045             await self._ensure_connected(timeout=timeout)
   1046         except OSError:
   1047             await self._close()

/anaconda/envs/azureml_custom_py37/lib/python3.7/site-packages/distributed/client.py in _ensure_connected(self, timeout)
   1098         try:
   1099             comm = await connect(
-> 1100                 self.scheduler.address, timeout=timeout, **self.connection_args
   1101             )
   1102             comm.name = "Client->Scheduler"

/anaconda/envs/azureml_custom_py37/lib/python3.7/site-packages/distributed/comm/core.py in connect(addr, timeout, deserialize, **connection_args)
    243                 backoff = min(backoff, 1)  # wait at most one second
    244             else:
--> 245                 _raise(error)
    246         else:
    247             break

/anaconda/envs/azureml_custom_py37/lib/python3.7/site-packages/distributed/comm/core.py in _raise(error)
    213             error,
    214         )
--> 215         raise IOError(msg)
    216 
    217     backoff = 0.01

OSError: Timed out trying to connect to 'tcp://10.233.0.7:8786' after 10 s: Timed out trying to connect to 'tcp://10.233.0.7:8786' after 10 s: in <distributed.comm.tcp.TCPConnector object at 0x7faaa01bca50>: ConnectionRefusedError: [Errno 111] Connection refused
```
​

**Comment by arnabbiswas1** [comment]
Looks like this entire problem is related to port forwarding. In last few days, I have tried to create a dask cluster in Azure ML workspace using cloud_provider from different VMs:

1. Running the code from the JupyterLab running on Compute Instance within the AzureML workspace. Ideally the daskcluster and the compute instance should be in the same subnet. But still the cluster creation failed.

2. Running the code from an Azure Data Science Virtual Machine within the IAAS infrastructure. Failed. Not sure if all the ports 9000, 9001, 9002 was open there. Didn't investigate.

3. Running the code from my laptop running on internet. It worked.

My code does the following:

a. Create AML cluster
b. print the AML Cluster variable.
c. Create a Dask client.

```
amlcluster = AzureMLCluster(ws,
                            vm_size="STANDARD_D1",
                            datastores=[Datastore.get(ws, "my_datastore")], 
                            environment_definition=ws.environments['AzureML-Dask-CPU'], 
                            initial_node_count=2, 
                            scheduler_idle_timeout=600,
                            vnet='some_vnet',
                            subnet='subnet1',
                            vnet_resource_group='some_rsrc_group',
                            ct_name="my_dask_cluster"
)


>> amlcluster

>> from dask.distributed import Client

c = Client(amlcluster)
c
```
In the local laptop, I didn't encounter the exception which I reported in this thread. However, I had to install JupyterLab extension:

```
conda install -c conda-forge nodejs
jupyter labextension install @jupyter-widgets/jupyterlab-manager
```

This is based on the suggestion provided in the [practicaldatascience](https://www.practicaldatascience.org/html/cloud_dask.html) page. I found the [documentation ](https://cloudprovider.dask.org/en/latest/azure.html) provided in the official documentation confusing.

Now, I really don't know why it didn't work  on Compute Instance or Data Science Virtual Machine. Considering Data Scientist community may not be familiar with port forwarding etc., it would be good if any assumptions/steps related to port forwarding is clearly documented.

**Comment by nickeubank** [comment]
I have three students who are getting the same thing, and we cannot for the life of us get it figured out. We've tried upgrading dask (to 2.30), dask-cloudprovider (0.4.1), upgrading jupyter, making sure widgets work, etc. 

When the students try and print `amlcluster`, they get: 

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
~/opt/anaconda3/lib/python3.7/site-packages/IPython/core/formatters.py in __call__(self, obj)
    916             method = get_real_method(obj, self.print_method)
    917             if method is not None:
--> 918                 method()
    919                 return True
    920 
~/opt/anaconda3/lib/python3.7/site-packages/distributed/deploy/cluster.py in _ipython_display_(self, **kwargs)
    339 
    340     def _ipython_display_(self, **kwargs):
--> 341         widget = self._widget()
    342         if widget is not None:
    343             return widget._ipython_display_(**kwargs)
~/opt/anaconda3/lib/python3.7/site-packages/dask_cloudprovider/providers/azure/azureml.py in _widget(self)
    858         jupyter = HTML(jupyter_link)
    859 
--> 860         status = HTML(self._widget_status(), layout=Layout(min_width="150px"))
    861 
    862         if self._supports_scaling:
~/opt/anaconda3/lib/python3.7/site-packages/dask_cloudprovider/providers/azure/azureml.py in _widget_status(self)
    765     def _widget_status(self):
    766         ### reporting proper number of nodes vs workers in a multi-GPU worker scenario
--> 767         nodes = len(self.scheduler_info["workers"])
    768 
    769         if self.use_gpu:
KeyError: 'workers'
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
~/opt/anaconda3/lib/python3.7/site-packages/IPython/core/formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)
--> 702             printer.pretty(obj)
    703             printer.flush()
    704             return stream.getvalue()
~/opt/anaconda3/lib/python3.7/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    403                         if cls is not object \
    404                                 and callable(cls.__dict__.get('__repr__')):
--> 405                             return _repr_pprint(obj, self, cycle)
    406 
    407             return _default_pprint(obj, self, cycle)
~/opt/anaconda3/lib/python3.7/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    693     """A pprint that just redirects to the normal repr function."""
    694     # Find newlines and replace them with p.break_()
--> 695     output = repr(obj)
    696     lines = output.splitlines()
    697     with p.group():
~/opt/anaconda3/lib/python3.7/site-packages/distributed/deploy/cluster.py in __repr__(self)
    367             self._cluster_class_name,
    368             self.scheduler_address,
--> 369             len(self.scheduler_info["workers"]),
    370             sum(w["nthreads"] for w in self.scheduler_info["workers"].values()),
    371         )
KeyError: 'workers'
```

And if they try and pass it to `Client()`, they get: 
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-7-3c5778c46797> in <module>
      1 from dask.distributed import Client
      2 
----> 3 c = Client(amlcluster)
~/opt/anaconda3/lib/python3.7/site-packages/distributed/client.py in __init__(self, address, loop, timeout, set_as_default, scheduler_file, security, asynchronous, name, heartbeat_interval, serializers, deserializers, extensions, direct_to_workers, **kwargs)
    721             ext(self)
    722 
--> 723         self.start(timeout=timeout)
    724         Client._instances.add(self)
    725 
~/opt/anaconda3/lib/python3.7/site-packages/distributed/client.py in start(self, **kwargs)
    894             self._started = asyncio.ensure_future(self._start(**kwargs))
    895         else:
--> 896             sync(self.loop, self._start, **kwargs)
    897 
    898     def __await__(self):
~/opt/anaconda3/lib/python3.7/site-packages/distributed/utils.py in sync(loop, func, callback_timeout, *args, **kwargs)
    346     if error[0]:
    347         typ, exc, tb = error[0]
--> 348         raise exc.with_traceback(tb)
    349     else:
    350         return result[0]
~/opt/anaconda3/lib/python3.7/site-packages/distributed/utils.py in f()
    330             if callback_timeout is not None:
    331                 future = asyncio.wait_for(future, callback_timeout)
--> 332             result[0] = yield future
    333         except Exception as exc:
    334             error[0] = sys.exc_info()
~/opt/anaconda3/lib/python3.7/site-packages/tornado/gen.py in run(self)
    733 
    734                     try:
--> 735                         value = future.result()
    736                     except Exception:
    737                         exc_info = sys.exc_info()
~/opt/anaconda3/lib/python3.7/site-packages/distributed/client.py in _start(self, timeout, **kwargs)
    989 
    990         try:
--> 991             await self._ensure_connected(timeout=timeout)
    992         except OSError:
    993             await self._close()
~/opt/anaconda3/lib/python3.7/site-packages/distributed/client.py in _ensure_connected(self, timeout)
   1072         else:
   1073             msg = await comm.read()
-> 1074         assert len(msg) == 1
   1075         assert msg[0]["op"] == "stream-start"
   1076 
AssertionError: 
```

Meanwhile, on Azure they seem to have three happy active nodes, so seems like the cluster was created, there's just a subsequent issue...

(@arnabbiswas1 you're making me feel bad because that's my site! And yet... I don't know the problem! Sorry. :( )


It's also extremely consistent -- we spent an hour and a half trying to figure it out, and over and over we got this.

**Comment by nickeubank** [comment]
@arnabbiswas1 what version of `distributed` do you have? Looks like my students had 2.11, and comparing 2.11 to 2.30, it looks like we might have an explanation: 

![image](https://user-images.githubusercontent.com/9683693/98851039-ab389700-2423-11eb-87aa-c98b1393a75b.png)

**Comment by nickeubank** [comment]
Replicated! I downgraded distributed to 2.11, and I got this:

![image](https://user-images.githubusercontent.com/9683693/98852424-a674e280-2425-11eb-8b75-f3e3b7e7d68d.png)

Opening new issue about bumping the version requirement.

**Comment by arnabbiswas1** [comment]
@nickeubank Not sure what is wrong here. My dask/distributed version is 2.20 in both the boxes where it's not working:

This is on the Azure IAAS VM:
````
dask                      2.20.0                     py_0
dask-cloudprovider        0.4.1                    pypi_0    pypi
dask-core                 2.20.0                     py_0
distributed               2.20.0                   py38_0
````

Azure Compute Instance:
````
dask                      2.20.0                     py_0
dask-cloudprovider        0.4.1                     <pip>
dask-core                 2.20.0                     py_0
distributed               2.20.0                   py37_0
````

**Comment by arnabbiswas1** [comment]
@nickeubank Thanks!  :clap:

I have updated the dask version to the latest (2.30) and :boom: that resolved the issue. I am able to create the client.

I am happy that it's not a port forwarding issue and simply a version mismatch.

**Comment by nickeubank** [comment]
Huh, ok! Guess the problem is more recent than 2.20! Maybe there's another line of code added somewhere elses. I'll change the bump suggestion in #165 to 2.30.

**Comment by jacobtomlinson** [comment]
Closed by #166
